lib/simplifyExpression/basicsSearch/removeAdditionOfZero.js

```diff
@@ -1,30 +1,32 @@
-const clone = require('../../util/clone');
+const {canApplyRule, applyRule} = require('math-rules');
 
 const ChangeTypes = require('../../ChangeTypes');
+const defineRuleString = require('../../util/defineRuleString');
 const Node = require('../../node');
 
+// e.g. 2 + 0 -> 2
+const REMOVE_ADDING_ZERO = defineRuleString('#a + 0', '#a');
+
+// e.g. 0 + 2 -> 2
+const REMOVE_ADDING_ZERO_REVERSE = defineRuleString('0 + #a', '#a');
+
 // If `node` is an addition node with 0 as one of its operands,
 // remove 0 from the operands list. Returns a Node.Status object.
 function removeAdditionOfZero(node) {
-  if (node.op !== '+') {
-    return Node.Status.noChange(node);
+  // TODO(kevinb): merge these two rules together
+  if (canApplyRule(REMOVE_ADDING_ZERO, node)) {
+    const newNode = applyRule(REMOVE_ADDING_ZERO, node);
+    return Node.Status.nodeChanged(
+      ChangeTypes.REMOVE_ADDING_ZERO, node, newNode);
   }
-  const zeroIndex = node.args.findIndex(arg => {
-    return Node.Type.isConstant(arg) && arg.value === '0';
-  });
-  let newNode = clone(node);
-  if (zeroIndex >= 0) {
-    // remove the 0 node
-    newNode.args.splice(zeroIndex, 1);
-    // if there's only one operand left, there's nothing left to add it to,
-    // so move it up the tree
-    if (newNode.args.length === 1) {
-      newNode = newNode.args[0];
-    }
+  else if (canApplyRule(REMOVE_ADDING_ZERO_REVERSE, node)) {
+    const newNode = applyRule(REMOVE_ADDING_ZERO_REVERSE, node);
```

Comment:
`canApplyRule` and `applyRule` both do their own searches so for more complicated expressions they'd return the root whereas the previous implementation would return false until `TreeSearch` finds the node that can actually be transformed in which case it returns that node.  I'm going to look at swapping out `TreeSearch` and figure out how to return the correct node.